RecoTracker/LSTCore/src/alpaka/Triplet.h

```diff
@@ -206,63 +206,304 @@ namespace lst {
+    short subdets = modulesInGPU.subdets[outerOuterLowerModuleIndex];
+
+    //calculate residual
+    if (layer3 <= 6 && ((side == lst::Center) or (drdz < 1))) { // for barrel
+      float paraA = r_init * r_init + 2 * (px * px + py * py) / (a * a) + 2 * (y_init * px - x_init * py) / a - r3 * r3;
+      float paraB = 2 * (x_init * px + y_init * py) / a;
+      float paraC = 2 * (y_init * px - x_init * py) / a + 2 * (px * px + py * py) / (a * a);
+      float A = paraB * paraB + paraC * paraC;
+      float B = 2 * paraA * paraB;
+      float C = paraA * paraA - paraC * paraC;
+      float sol1 = (-B + alpaka::math::sqrt(acc, B * B - 4 * A * C)) / (2 * A);
+      float sol2 = (-B - alpaka::math::sqrt(acc, B * B - 4 * A * C)) / (2 * A);
+      float solz1 = alpaka::math::asin(acc, sol1) / rou * pz / p + z_init;
+      float solz2 = alpaka::math::asin(acc, sol2) / rou * pz / p + z_init;
+      float diffz1 = (solz1 - z3) * 100;
+      float diffz2 = (solz2 - z3) * 100;
+      if (alpaka::math::isnan(acc, diffz1))
+        residual = diffz2;
+      else if (alpaka::math::isnan(acc, diffz2))
+        residual = diffz1;
+      else {
+        residual = (alpaka::math::abs(acc, diffz1) < alpaka::math::abs(acc, diffz2)) ? diffz1 : diffz2;
+      }
+    } else { // for endcap
+      float s = (z_target - z_init) * p / pz;
+      float x = x_init + px / a * alpaka::math::sin(acc, rou * s) - py / a * (1 - alpaka::math::cos(acc, rou * s));
+      float y = y_init + py / a * alpaka::math::sin(acc, rou * s) + px / a * (1 - alpaka::math::cos(acc, rou * s));
+      residual = (r_target - alpaka::math::sqrt(acc, x * x + y * y)) * 100; 
+    }
+    
+    // error
+    if (moduleType3 == 0) { 
+      error = 0.15f;        //PS
+    } else  {              
+      error = 5.0f;         //2S
+    }
+
+    float projection_missing2 = 1;
+    if (drdz < 1)
+      projection_missing2 = ((subdets == lst::Endcap) or (side == lst::Center))
+                                ? 1.f
+                                : 1 / (1 + drdz * drdz);  // cos(atan(drdz)), if dr/dz<1
+    if (drdz > 1)
+      projection_missing2 = ((subdets == lst::Endcap) or (side == lst::Center))
+                                ? 1.f
+                                : drdz * drdz / (1 + drdz * drdz);  //sin(atan(drdz)), if dr/dz>1
+
+    rzChiSquared = 12 * (residual * residual) / (error * error * projection_missing2);
+
+    //helix calculation returns NaN, use linear approximation
+    if (alpaka::math::isnan(acc, rzChiSquared) || circleRadius < 0) {
+      float slope = (z_other - z1) / (r_other - r1);
+
+      residual = (layer3 <= 6) ? ((z_target - z1) - slope * (r_target - r1)) : ((r_target - r1) - (z_target - z1) / slope);
+      residual = (moduleType3 == 0) ? residual / 0.15f : residual / 5.0f;
+
+      rzChiSquared = 12 * residual * residual;
+      return rzChiSquared < 2.7711823e-4;
     }
-  };
+
+    //cuts for different regions
```

Comment:
I've added a link to this PR, since the past presentations are outdated

Reply:
I don't think this is enough: this PR contains links to personal web pages, which may be deleted some time from now. On the other hand, files uploaded to indico agendas are "more guaranteed" to stay there for longer. We have lost a lot of documentation by relying on personal web pages, so I don't think we should keep doing it. Therefore, I would have to insist to have a presentation with the results (nothing too complicated but some decent documentation) uploaded in some indico agenda, and then link that to this PR and potentially in the code.

Reply:
If it helps, I added a pdf link matching the gslides shown on Aug 30 https://indico.cern.ch/event/1441459/#12-contribution

For the links to validation plots in the PR description, I can suggest to open the links and copy-paste e.g. plots vs eta in the PR description itself. OTOH, most of these plots illustrate what can be seen from the CI plots available in this PR. So, it will not really add more info than what's already available

Reply:
Thanks, Slava! I added the link the in the PR description, and I am resolving this comment thread.